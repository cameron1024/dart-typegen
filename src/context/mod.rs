use std::path::{Path, PathBuf};

use miette::{Diagnostic, IntoDiagnostic, Result};
use thiserror::Error;

use crate::model::Library;

pub struct Context {
    pub path: PathBuf,
    pub text: String,
    pub library: Library,
}

impl Context {
    pub fn from_memory(text: &str) -> Result<Self> {
        todo!()
    }

    pub fn from_path(path: &Path) -> Result<Self> {
        let text = std::fs::read_to_string(path).into_diagnostic()?;
        let parsed = Library::parse_file(path)?;

        Ok(Context {
            path: path.to_path_buf(),
            text,
            library: parsed,
        })
    }

    pub fn resolve_path(&self, path: &Path) -> Result<PathBuf, ResolvePathError> {
        // unwrap is fine because all parent-less paths would have failed to parse the config
        let parent = self.path.parent().unwrap();

        parent
            .join(path)
            .canonicalize()
            .map_err(|io| ResolvePathError {
                path: path.to_path_buf(),
                dir: if parent == Path::new("") {
                    std::fs::canonicalize(".")
                } else {
                    parent.canonicalize()
                }
                .unwrap(),
                io,
            })
    }
}

#[derive(Debug, Error, Diagnostic)]
#[error("Failed to resolve the path `{path}` relative to `{dir}`")]
pub struct ResolvePathError {
    path: PathBuf,
    dir: PathBuf,

    #[source]
    io: std::io::Error,
}
