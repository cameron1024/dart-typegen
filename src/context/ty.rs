use std::{borrow::Cow, fmt::Display};

use chumsky::{
    extra::Err,
    label::LabelError,
    prelude::*,
    text::{Char, TextExpected},
    util::MaybeRef,
};
use knus::span::Span;
use miette::{Diagnostic, NamedSource, SourceSpan};
use thiserror::Error;

use crate::model::SpannedScalar;

use super::*;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Ty {
    // Span relative to document, the type string
    pub span: Span,
    pub kind: TyKind,
}

impl Display for Ty {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.kind {
            TyKind::Simple(ident) => write!(f, "{ident}"),
            TyKind::List(inner) => write!(f, "List<{inner}>"),
            TyKind::Set(inner) => write!(f, "Set<{inner}>"),
            TyKind::Map { key, value } => write!(f, "Map<{key}, {value}>"),
            TyKind::Nullable(inner) => write!(f, "{inner}?"),
        }
    }
}

impl Context {
    pub fn parse_ty(&self, value: &SpannedScalar<String>) -> (Option<Ty>, Vec<miette::Report>) {
        let span_offset = value.span.offset();
        let (output, errors) = ty(span_offset).parse(value).into_output_errors();

        let errors = errors.into_iter().map(|err| ParseDartTypeError {
            src: self.named_source(),
            span: knus::span::Span(err.span().start + span_offset, err.span().end + span_offset)
                .into(),
            reason: err.reason().to_string(),
        });

        (output, errors.map(Into::into).collect())
    }

    pub fn named_source(&self) -> NamedSource<String> {
        let source_name = match &self.path {
            Some(path) => path.to_string_lossy(),
            None => Cow::Borrowed("<memory>"),
        };
        NamedSource::new(source_name, self.text.clone()).with_language("kdl")
    }
}

#[derive(Debug, Error, Diagnostic)]
#[error("Failed to parse Dart type")]
struct ParseDartTypeError {
    #[source_code]
    src: NamedSource<String>,

    #[label]
    span: SourceSpan,

    #[help]
    reason: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TyKind {
    Simple(String),
    Nullable(Box<Ty>),
    List(Box<Ty>),
    Set(Box<Ty>),
    Map { key: Box<Ty>, value: Box<Ty> },
}

fn ty<'a>(span_offset: usize) -> impl Parser<'a, &'a str, Ty, Err<Rich<'a, char>>> {
    recursive(move |ty| {
        let list = just("List")
            .then_ignore(whitespace().repeated())
            .then_ignore(just('<'))
            .then_ignore(whitespace().repeated())
            .then(ty.clone())
            .then_ignore(whitespace().repeated())
            .then_ignore(just(',').or_not())
            .then_ignore(whitespace().repeated())
            .then_ignore(just('>'))
            .map(|(_list, ty)| TyKind::List(Box::new(ty)));

        let set = just("Set")
            .then_ignore(whitespace().repeated())
            .then_ignore(just('<'))
            .then_ignore(whitespace().repeated())
            .then(ty.clone())
            .then_ignore(whitespace().repeated())
            .then_ignore(just(',').or_not())
            .then_ignore(whitespace().repeated())
            .then_ignore(just('>'))
            .map(|(_, ty)| TyKind::Set(Box::new(ty)));

        let map = just("Map")
            .then_ignore(whitespace().repeated())
            .then_ignore(just('<'))
            .then_ignore(whitespace().repeated())
            .then(ty.clone())
            .then_ignore(whitespace().repeated())
            .then_ignore(just(','))
            .then_ignore(whitespace().repeated())
            .then(ty.clone())
            .then_ignore(whitespace().repeated())
            .then_ignore(just(',').or_not())
            .then_ignore(whitespace().repeated())
            .then_ignore(just('>'))
            .map(|((_, key), value)| TyKind::Map {
                key: Box::new(key),
                value: Box::new(value),
            });

        let simple = ident().map(|s| TyKind::Simple(s.to_string()));

        let all = choice((list, set, map, simple)).map_with(move |kind, extra| {
            let span = extra.span();
            Ty {
                kind,
                span: Span(span.start + span_offset, span.end + span_offset),
            }
        });
        let nullable =
            all.then(just('?').or_not())
                .map_with(move |(inner, huh), extra| match huh {
                    None => inner,
                    Some(_) => {
                        let span = extra.span();
                        Ty {
                            kind: TyKind::Nullable(Box::new(inner)),
                            span: Span(span.start + span_offset, span.end + span_offset),
                        }
                    }
                });

        whitespace()
            .repeated()
            .ignore_then(nullable)
            .then_ignore(whitespace().repeated())
    })
}
fn whitespace<'a>() -> impl Parser<'a, &'a str, (), Err<Rich<'a, char>>> + Clone {
    choice((just(' '), just('\n'), just('\t'))).ignored()
}

/// a modified version of [`chumsky::text::ident()`] that accepts $ as a valid char
/// anywhere
fn ident<'a>() -> impl Parser<'a, &'a str, &'a str, Err<Rich<'a, char>>> + Clone {
    any()
        .try_map(|c: char, span| {
            if c.is_ident_start() || c == '$' {
                Ok(c)
            } else {
                Err(LabelError::<&'a str, _>::expected_found(
                    [TextExpected::<&'a str>::IdentifierPart],
                    Some(MaybeRef::Val(c)),
                    span,
                ))
            }
        })
        .then(
            any()
                .try_map(|c: char, span| {
                    if c.is_ident_continue() || c == '$' {
                        Ok(c)
                    } else {
                        Err(LabelError::<&'a str, _>::expected_found(
                            [TextExpected::<&'a str>::IdentifierPart],
                            Some(MaybeRef::Val(c)),
                            span,
                        ))
                    }
                })
                .repeated(),
        )
        .to_slice()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse(src: &str) -> Ty {
        ty(0).parse(src).unwrap()
    }

    #[test]
    fn can_parse_types() {
        let _ = parse("Asdf");
        let _ = parse("asdf");
        let _ = parse("$asdf");
        let _ = parse("     $_123asdf     ");
        let _ = parse("List<A,>");
        let _ = parse("List    <    A    ,      >");
        let _ = parse("Map<A, B>");

        let ty = parse("Hello");
        assert_eq!(ty.kind, TyKind::Simple("Hello".to_string()));

        let ty = parse("$_Hello123$");
        assert_eq!(ty.kind, TyKind::Simple("$_Hello123$".to_string()));

        let ty = parse("Hello?");
        assert!(matches!(ty.kind, TyKind::Nullable(_)));

        // while we probably want to error here, the parser should't throw a fit if it encounters
        // this
        let ty = parse("List");
        assert_eq!(ty.kind, TyKind::Simple("List".to_string()));

        let ty = parse("Set");
        assert_eq!(ty.kind, TyKind::Simple("Set".to_string()));

        let ty = parse("Map");
        assert_eq!(ty.kind, TyKind::Simple("Map".to_string()));

        // collections
        let ty = parse("List<Hello>");
        assert!(
            matches!(ty.kind, TyKind::List(inner) if inner.kind == TyKind::Simple("Hello".to_string()))
        );
        let ty = parse("   List<   Hello    >       ");
        assert!(
            matches!(ty.kind, TyKind::List(inner) if inner.kind == TyKind::Simple("Hello".to_string()))
        );

        let ty = parse("Set<Hello,>");
        assert!(
            matches!(ty.kind, TyKind::Set(inner) if inner.kind == TyKind::Simple("Hello".to_string()))
        );

        let ty = parse("Map<$,_$123,>");
        assert!(matches!(
                ty.kind, 
                TyKind::Map {
                    key,
                    value,
                } if key.kind == TyKind::Simple("$".to_string()) 
                  && value.kind == TyKind::Simple("_$123".to_string())));
    }
}
