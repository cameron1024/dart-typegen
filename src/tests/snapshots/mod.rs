use crate::context::Context;

macro_rules! output_snapshot {
    ($name:ident) => {
        #[test]
        fn $name() {
            let context = Context::from_str(include_str!(crate::test_file!($name))).unwrap();
            let output = context.codegen_to_string().unwrap();

            insta::assert_snapshot!(stringify!($name), output, "dart");
        }
    };
}

crate::all_test_files!(output_snapshot);
