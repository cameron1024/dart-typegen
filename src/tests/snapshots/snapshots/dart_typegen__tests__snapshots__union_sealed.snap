---
source: src/tests/snapshots/mod.rs
expression: dart
---
import "package:equatable/equatable.dart";

sealed class Animal with EquatableMixin {
  const Animal();

  Map<String, dynamic> toJson();
  factory Animal.fromJson(Map<String, dynamic> json) => switch (json["type"]) {
    "Dog" => Dog.fromJson(json),
    "Cat" => Cat.fromJson(json),
    final other => throw ArgumentError("unknown discriminant: $other"),
  };
}

final class Dog extends Animal with EquatableMixin {
  final String name;

  const Dog({required this.name}) : super();

  @override
  List<Object?> get props => [name];

  DogBuilder toBuilder() => DogBuilder(name: name);

  @override
  Map<String, dynamic> toJson() => {"name": name, "type": "Dog"};
  factory Dog.fromJson(Map<String, dynamic> json) =>
      Dog(name: json["name"] as String);
}

/// Builder class for [Dog]
final class DogBuilder {
  String name;

  DogBuilder({required this.name});

  Dog build() => Dog(name: name);
}

final class Cat extends Animal with EquatableMixin {
  final int age;

  const Cat({required this.age}) : super();

  @override
  List<Object?> get props => [age];

  CatBuilder toBuilder() => CatBuilder(age: age);

  @override
  Map<String, dynamic> toJson() => {"age": age, "type": "Cat"};
  factory Cat.fromJson(Map<String, dynamic> json) =>
      Cat(age: json["age"] as int);
}

/// Builder class for [Cat]
final class CatBuilder {
  int age;

  CatBuilder({required this.age});

  Cat build() => Cat(age: age);
}
